src/common/src/util/value_encoding/column_aware_row_encoding.rs

```diff
@@ -105,8 +106,77 @@ impl RowEncoding {
     }
 }
 
+trait ValueRowSerializer: Clone {
+    fn new(column_ids: &[ColumnId]) -> impl ValueRowSerializer;
+    fn serialize(&self, row: impl Row) -> Vec<u8>;
+}
+
+trait ValueRowDeserializer: Clone {
+    fn new(column_ids: &[ColumnId], schema: Arc<[DataType]>) -> impl ValueRowDeserializer;
+    fn deserialize(&self, encoded_bytes: &[u8]) -> Result<Vec<Datum>>;
+}
+
+pub trait ValueRowSerde: Clone {
+    fn new(column_ids: &[ColumnId], schema: Arc<[DataType]>) -> Self;
+    fn serialize(&self, row: impl Row) -> Vec<u8>;
+    fn deserialize(&self, encoded_bytes: &[u8]) -> Result<Vec<Datum>>;
+}
+
+#[derive(Clone)]
+pub struct BasicSerializer {}
+
+impl ValueRowSerializer for BasicSerializer {
+    fn new(_column_ids: &[ColumnId]) -> BasicSerializer {
+        BasicSerializer {}
+    }
+
+    fn serialize(&self, row: impl Row) -> Vec<u8> {
+        let mut buf = vec![];
+        for datum in row.iter() {
+            serialize_datum_into(datum, &mut buf);
+        }
+        buf
+    }
+}
+
+impl ValueRowDeserializer for RowDeserializer {
+    fn new(_column_ids: &[ColumnId], schema: Arc<[DataType]>) -> RowDeserializer {
+        RowDeserializer::new(schema.as_ref().to_owned())
+    }
+
+    fn deserialize(&self, encoded_bytes: &[u8]) -> Result<Vec<Datum>> {
+        Ok(self.deserialize(encoded_bytes)?.into_inner())
+    }
+}
+
+#[derive(Clone)]
+pub struct BasicSerde {
+    deserializer: RowDeserializer,
+}
+
+impl ValueRowSerde for BasicSerde {
+    fn new(column_ids: &[ColumnId], schema: Arc<[DataType]>) -> BasicSerde {
+        BasicSerde {
+            deserializer: <RowDeserializer as ValueRowDeserializer>::new(column_ids, schema),
+        }
+    }
+
+    fn serialize(&self, row: impl Row) -> Vec<u8> {
+        let mut buf = vec![];
+        for datum in row.iter() {
+            serialize_datum_into(datum, &mut buf);
+        }
+        buf
+    }
```

Comment:
Since there is a `struct BasicSerializer` above, I think it's better to have a field `serializer: BasicSerializer` in `BasicSerde`, and do `self.serializer.serialize` in impl of `ValueRowSerde::serialize`.

Reply:
ah, I forgot to use it as I wrote it a long time ago😅